linkml_runtime/utils/schemaview.py

```diff
@@ -92,13 +91,90 @@ class OrderedBy(Enum):
     """
 
 
+WHITE = 0
+GREY = 1
+BLACK = 2
+
+
+def detect_cycles(f: Callable[[Any], Iterable[Any] | None], x: Any) -> None:
+    """Detect cycles in a graph, using function `f` to walk the graph, starting at node `x`.
+
+    Uses the classic white/grey/black colour coding algorithm to track which nodes have been explored. In this
+    case, "node" refers to any element in a schema and "neighbours" are elements that can be reached from that
+    node by executing function `f`.
+
+    WHITE: unexplored
+    GREY:  node is being processed; processing includes exploring all neighbours reachable via f(node)
+    BLACK: node and all of its neighbours (and their neighbours, etc.) have been processed
```

Comment:
This naming system could be changed (e.g. to `TODO` / `PROCESSING` / `DONE` or something similar) but the white/grey/black nomenclature is used in numerous places where this algorithm is described, so may as well stick with it.